galasa-managers-parent/galasa-managers-comms-parent/dev.galasa.mq.manager/src/main/java/dev/galasa/mq/internal/MQManagerImpl.java

```diff
@@ -0,0 +1,258 @@
+import dev.galasa.mq.Queue;
+import dev.galasa.mq.QueueManager;
+import dev.galasa.mq.internal.properties.InstanceChannelName;
+import dev.galasa.mq.internal.properties.InstanceCredentials;
+import dev.galasa.mq.internal.properties.InstanceForTag;
+import dev.galasa.mq.internal.properties.InstanceHost;
+import dev.galasa.mq.internal.properties.MqPropertiesSingleton;
+import dev.galasa.mq.internal.properties.QueueNameForTag;
+import dev.galasa.mq.internal.properties.InstanceName;
+import dev.galasa.mq.internal.properties.InstancePort;
+
+@Component(service = { IManager.class })
+public class MQManagerImpl extends AbstractManager {
+
+    private static final Log  logger = LogFactory.getLog(MQManagerImpl.class);
+    private static final String NAMESPACE   = "mq";
+    
+    private Path storedArtifactsRoot;
+    private ICredentialsService credentialService;
+    
+    private HashMap<String,MessageQueueManagerImpl> queueManagers = new HashMap<>();
+    private List<MessageQueueImpl> queues = new ArrayList<>();
+    
+    private String currentMethod = new String();
+
+    @GenerateAnnotatedField(annotation = Queue.class)
+    public IMessageQueue generateMessageQueue(Field field, List<Annotation> annotations) throws MqManagerException {
+    	Queue annotation = field.getAnnotation(Queue.class);
+    	
+    	//check that we have provisioned the qmgr for this queue
+    	String qmgrTag = annotation.queueMgrTag();
+    	if(!queueManagers.containsKey(qmgrTag)) {
+    		throw new MqManagerException("Unable to provision queue: " + annotation.name() + " no QMGR found tagged: " + qmgrTag);
+    	}
+    	
+    	//check if the name is specified in the queue or in the CPS
+    	String queueName = annotation.name();
+    	String queueTag = annotation.tag();
+    	
+    	if(queueName.isEmpty() && queueTag.isEmpty())
+    		throw new MqManagerException("Either name or tag must be specified in @Queue annotation");
+    	
+    	if(!queueName.isEmpty() && !queueTag.isEmpty())
+    		throw new MqManagerException("Both name and tag are specified in @Queue annotation, these are mutually exclusive");
+    	
+    	String name = queueName;
+    	if(queueTag.length() > 0){
+    		name = QueueNameForTag.get(queueTag);
+    		if(name == null) {
+    			throw new MqManagerException("No queue tag for " + queueTag);
+    		}
+    	}
+    	
+    	//construct the queue and add it to our list
+        MessageQueueImpl queue = new MessageQueueImpl(name,queueManagers.get(qmgrTag), annotation.archive(), this);
+        this.queues.add(queue);
+        registerAnnotatedField(field, queue);
+        return queue;
+    }
+    
+    @GenerateAnnotatedField(annotation = QueueManager.class)
+    public IMessageQueueManager generateMessageQueueManager(Field field, List<Annotation> annotations) throws MqManagerException{
+    	//obtain the tag for this queue manager
+    	QueueManager annotation = field.getAnnotation(QueueManager.class);
+    	String tag = annotation.queueMgrTag();
+		
```

Comment:
should be checking this.queueManagers to see if you have already provisioned the qmgr,     just in case there are 2 fields with the same tag